types/wlr_linux_dmabuf.c

```diff
@@ -0,0 +1,449 @@
+	return dmabuf->attributes.flags
+		& ZWP_LINUX_BUFFER_PARAMS_V1_FLAGS_Y_INVERT;
+}
+
+struct wlr_dmabuf_buffer *wlr_dmabuf_buffer_from_buffer_resource(
+		struct wl_resource *buffer_resource) {
+	assert(wl_resource_instance_of(buffer_resource, &wl_buffer_interface,
+			&wl_buffer_impl));
+
+	struct wlr_dmabuf_buffer *buffer = wl_resource_get_user_data(buffer_resource);
+	assert(buffer);
+	assert(buffer->buffer_resource);
+	assert(!buffer->params_resource);
+	assert(buffer->buffer_resource == buffer_resource);
+
+	return buffer;
+}
+
+static void linux_dmabuf_buffer_destroy(struct wlr_dmabuf_buffer *buffer) {
+	for (int i = 0; i < buffer->attributes.n_planes; i++) {
+		close(buffer->attributes.fd[i]);
+		buffer->attributes.fd[i] = -1;
+	}
+	buffer->attributes.n_planes = 0;
+	free(buffer);
+}
+
+static void params_destroy(struct wl_client *client, struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static void params_add(struct wl_client *client,
+		struct wl_resource *params_resource, int32_t name_fd,
+		uint32_t plane_idx, uint32_t offset, uint32_t stride,
+		uint32_t modifier_hi, uint32_t modifier_lo) {
+	struct wlr_dmabuf_buffer *buffer = wlr_dmabuf_buffer_from_params_resource(
+		params_resource);
+
+	if (!buffer) {
+		wl_resource_post_error(params_resource,
+			ZWP_LINUX_BUFFER_PARAMS_V1_ERROR_ALREADY_USED,
+			"params was already used to create a wl_buffer");
+		close(name_fd);
+		return;
+	}
+
+	if (plane_idx >= WLR_LINUX_DMABUF_MAX_PLANES) {
+		wl_resource_post_error(params_resource,
+			ZWP_LINUX_BUFFER_PARAMS_V1_ERROR_PLANE_IDX,
+			"plane index %u > %u", plane_idx, WLR_LINUX_DMABUF_MAX_PLANES);
+		close(name_fd);
+		return;
+	}
+
+	if (buffer->attributes.fd[plane_idx] != -1) {
+		wl_resource_post_error(params_resource,
+			ZWP_LINUX_BUFFER_PARAMS_V1_ERROR_PLANE_SET,
+			"a dmabuf with id %d has already been added for plane %u",
+			buffer->attributes.fd[plane_idx],
+			plane_idx);
+		close(name_fd);
+		return;
+	}
+
+	buffer->attributes.fd[plane_idx] = name_fd;
+	buffer->attributes.offset[plane_idx] = offset;
+	buffer->attributes.stride[plane_idx] = stride;
+	buffer->attributes.modifier[plane_idx] = ((uint64_t)modifier_hi << 32) |
+		modifier_lo;
+	buffer->attributes.n_planes++;
+}
+
+static void handle_buffer_destroy(struct wl_resource *buffer_resource)
+{
+	struct wlr_dmabuf_buffer *buffer = wlr_dmabuf_buffer_from_buffer_resource(
+		buffer_resource);
+
+	linux_dmabuf_buffer_destroy(buffer);
+}
+
+static void params_create_common(struct wl_client *client,
+		struct wl_resource *params_resource, uint32_t buffer_id, int32_t width,
+		int32_t height, uint32_t format, uint32_t flags) {
+	struct wlr_dmabuf_buffer *buffer = wlr_dmabuf_buffer_from_params_resource(
+		params_resource);
+	if (!buffer) {
+		wl_resource_post_error(params_resource,
+			ZWP_LINUX_BUFFER_PARAMS_V1_ERROR_ALREADY_USED,
+			"params was already used to create a wl_buffer");
+		return;
```

Comment:
This branch is never reached, because `wlr_dmabuf_buffer_from_params_resource` asserts that the buffer isn't NULL.

A fix would be to replace the condition by `if (!wl_resource_get_user_data(params_resource)) {` and move the `wlr_dmabuf_buffer_from_params_resource` call below.